src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,529 @@
+            async with aiohttp.ClientSession() as session:
+                headers = {
+                    "accept": "application/json",
+                    "Content-Type": "application/json",
+                }
+                async with session.post(
+                    f"{server}/{self.worker_url}",
+                    json=batch_request_data.dict(),
+                    timeout=self._timeout_inference_request,
+                    headers=headers,
+                ) as response:
+                    if response.status == 408:
+                        raise HTTPException(408, "Request timed out")
+                    response.raise_for_status()
+                    response = await response.json()
+                    outputs = response["outputs"]
+                    assert len(batch) == len(outputs), f"result has {len(outputs)} items but batch is {len(batch)}"
+                    result = {request[0]: r for request, r in zip(batch, outputs)}
+                    self._responses.update(result)
+        except Exception as e:
+            result = {request[0]: e for request in batch}
+            self._responses.update(result)
+
+    async def consumer(self):
+        while True:
+            await asyncio.sleep(0.05)
+
+            has_sent = False
+
+            batch = self._batch[: self.max_batch_size]
+            while batch and (
+                (len(batch) >= self.max_batch_size) or ((time.time() - self._last_batch_sent) > self.timeout_batch)
+            ):
+                has_sent = True
+
+                asyncio.create_task(self.send_batch(batch))
+
+                self._batch = self._batch[self.max_batch_size :]
+                batch = self._batch[: self.max_batch_size]
+
+            if has_sent:
+                self._last_batch_sent = time.time()
+
+    async def process_request(self, data: BaseModel):
+        if not self.servers:
+            raise HTTPException(500, "None of the workers are healthy!")
+
+        request_id = uuid.uuid4().hex
+        request: Tuple = (request_id, data)
+        self._batch.append(request)
+
+        while True:
+            await asyncio.sleep(0.05)
+
+            if request_id in self._responses:
+                result = self._responses[request_id]
+                del self._responses[request_id]
+                _raise_granular_exception(result)
+                return result
+
+    def run(self):
+        if self._server_ready:
+            return
+
+        INPUT_SCHEMA = self._input_schema
+        OUTPUT_SCHEMA = self._output_schema
+
+        logger.info(f"servers: {self.servers}")
+
+        self._ITER = cycle(self.servers)
+        self._last_batch_sent = time.time()
+
+        fastapi_app = _create_fastapi("Load Balancer")
+        security = HTTPBasic()
+        fastapi_app.global_request_count = 0
+        fastapi_app.num_current_requests = 0
+        fastapi_app.last_process_time = 0
+        fastapi_app.SEND_TASK = None
+
+        @fastapi_app.on_event("startup")
```

Comment:
do we really need to define all the local functions inside `run`? can it be moved to class as static if needed and just called here...

Reply:
Let me see if I can make `run` more readable.